boring/src/ssl/test/mod.rs

```diff
@@ -480,6 +480,76 @@ fn test_alpn_server_unilateral() {
     assert_eq!(None, s.ssl().selected_alpn_protocol());
 }
 
+#[test]
+fn test_select_cert_ok() {
+    let mut server = Server::builder();
+    server
+        .ctx()
+        .set_select_certificate_callback(|_client_hello| Ok(()));
+    let server = server.build();
+
+    let client = server.client();
+    client.connect();
+}
+
+#[test]
+fn test_select_cert_error() {
+    let mut server = Server::builder();
+    server.should_error();
+    server
+        .ctx()
+        .set_select_certificate_callback(|_client_hello| Err(ssl::SelectCertError::ERROR));
+    let server = server.build();
+
+    let client = server.client();
+    client.connect_err();
+}
+
+#[test]
+fn test_select_cert_unknown_extension() {
+    let mut server = Server::builder();
+    let unknown_extension = std::sync::Arc::new(std::sync::Mutex::new(None));
+    server.ctx().set_select_certificate_callback({
+        let unknown = unknown_extension.clone();
+        move |client_hello| {
+            *unknown.lock().unwrap() = client_hello.get_extension(1337).map(ToOwned::to_owned);
+            Ok(())
+        }
+    });
+    let server = server.build();
+
+    let client = server.client();
+    client.connect();
+    assert_eq!(unknown_extension.lock().unwrap().as_deref(), None);
+}
+
+#[test]
+fn test_select_cert_alpn_extension() {
+    let mut server = Server::builder();
+    let alpn_extension = std::sync::Arc::new(std::sync::Mutex::new(None));
+    server.ctx().set_select_certificate_callback({
+        let alpn = alpn_extension.clone();
+        move |client_hello| {
+            *alpn.lock().unwrap() = Some(
+                client_hello
+                    .get_extension(ffi::TLSEXT_TYPE_application_layer_protocol_negotiation as u16)
+                    .unwrap()
+                    .to_owned(),
+            );
+            Ok(())
+        }
+    });
+    let server = server.build();
+
+    let mut client = server.client();
+    client.ctx().set_alpn_protos(b"\x06http/2").unwrap();
+    client.connect();
+    assert_eq!(
+        alpn_extension.lock().unwrap().as_deref(),
+        Some(&b"\x00\x07\x06http/2"[..])
```

Comment:
Note that it returns the entire extension data, including the leading length, not sure that's good or not.

Reply:
Yeah, better strip the leading length

Reply:
I've stripped the length and made the function panic if that doesn't correspond to the length of the rest of the slice. AFAIK the handshake wouldn't have reached the point where you can call that function if the lengths didn't correspond.